Reorganize imgpkg Docs, Write Workflow Docs, and Add imgpkg tag list Docs (#21)
carvel-dev/imgpkg: docs/basic-workflow.md

#### How to distribute the image

In the folder [examples/basic](../examples/basic), there is a set of configuration files that
will allow a user to create a service and a deployment of a simple application that will run on 
Kubernetes.

```
examples/basic/
├── deployment.yml
└── service.yml
```

You can push the above folder to a local Docker registry using the following command:

`imgpkg push -f examples/basic -i localhost:9001/simple-app-configuration`

Flags used in the command:
  * `-f` indicates the folder the user wants to package as an OCI Image
  * `-i` indicates that the user want to push a simple OCI image to a registry

The output will display all the files that will be packaged and the destination of the image:
```
dir: .
file: deployment.yml
file: service.yml
Pushed 'localhost:9001/simple-app-configuration@sha256:98ff397d8a8200ecb228c9add5767ef40c4e59d751a6e85880a1f903394ee3e7'
Succeeded
```

#### How to retrieve the image

You can run the following command to download the configuration:

`imgpkg pull -o /tmp/simple-app-config -i localhost:9001/simple-app-configuration`

Flags used in the command:
  * `-o` indicates the folder where the OCI image will be unpacked

>   * `-i` indicates that the user want to pull a simple OCI Image from the Registry

Reviewer: ```suggestion
  * `-i` indicates that the user wants to pull a simple OCI Image from the Registry
```

8:40:24AM: ---- applying complete [2/2 done] ----
8:40:24AM: ---- waiting complete [2/2 done] ----

Succeeded
```

You have successfully deployed an OCI image to Kubernetes using `imgpkg` and `kapp`. Before continuing, 
remove the application from your cluster:

```shell
kapp delete -a simple-app -y
```

### Bundle distribution

For more information on bundles, read more [here](README.md#images-vs-bundles).

#### How to distribute the bundle

In the folder [examples/basic-bundle](../examples/basic-bundle), there is a set of configuration files that
will allow a user to create a service and a deployment for an application that will run on Kubernetes. The 
folder also contains the [`.imgpkg`](resources.md#imageslock) hidden directory with an `ImagesLock` file and 
a [bundle metadata file](resources.md#bundle-metadata).

```shell
examples/basic-bundle
├── .imgpkg
│   ├── bundle.yml
│   └── images.yml
└── config.yml
```

You can push the above folder containing a bundle to your local Docker registry using the following command:

`imgpkg push -f examples/basic-bundle -b localhost:5000/simple-app-bundle`

Flags used in the command:

>   * `-f` indicates the folder the user want to package

Reviewer: ```suggestion
  * `-f` indicates the folder the user wants to package
```

>   * `-b` indicates that the user want to push a bundle to the registry

Reviewer: ```suggestion
  * `-b` indicates that the user wants to push a bundle to the registry
```

The output displays all the files that will be packaged and the destination of the bundle:
```shell
dir: .
dir: .imgpkg
file: .imgpkg/bundle.yml
file: .imgpkg/images.yml
file: config.yml
Pushed 'localhost:5000/simple-app-bundle@sha256:ec3f870e958e404476b9ec67f28c598fa8f00f819b8ae05ee80d51bac9f35f5d'
Succeeded
```

#### How to retrieve the bundle

You can retrieve the bundle by running the following command to download the bundle:

`imgpkg pull -o /tmp/simple-app-bundle -b localhost:5000/simple-app-bundle`

Flags used in the command:
  * `-o` indicates the folder where the bundle will be unpacked

>   * `-b` indicates that the user want to pull a bundle from the registry

Reviewer: ```suggestion
  * `-b` indicates that the user wants to pull a bundle from the registry
```

The output shows the image pull was successful:
```shell
Pulling image 'localhost:5000/simple-app-bundle@sha256:ec3f870e958e404476b9ec67f28c598fa8f00f819b8ae05ee80d51bac9f35f5d'
Extracting layer 'sha256:7906b9650be657359ead106e354f2728e16c8f317e1d87f72b05b5c5ec3d89cc' (1/1)
Locating image lock file images...
One or more images not found in bundle repo; skipping lock file update

Succeeded
```

__Note:__ The message `One or more images not found in bundle repo; skipping lock file update` indicates that the file 

> `/tmp/simple-app-bundle/.imgpkg/images.yml` was not updated since the bundle has never been pushed to the Docker registry 

Reviewer: ```suggestion
`/tmp/simple-app-bundle/.imgpkg/images.yml` was not updated since the bundle has never been copied to a different Docker registry 
```

#### Scenario

In this scenario, an application developer creates a bundle with images and configurations and another develop that wants
to use the application prefers to use a different registry to store the images and configuration. This scenario can happen 
if you are downloading an application from an external source and want to use a registry that is associated with the Kubernetes deployment.

__Note:__ The example below shows how a bundle is being relocated, but the same flow should apply for relocating non bundle images. Simply changing the `-b` option to `-i` and using the folder in the [Image distribution](#image-distribution) scenario will showcase the same concepts using `imgpkg`.

#### How to relocate bundle

In the folder [examples/basic-bundle](../examples/basic-bundle) there is a set of configuration files that
will allow a user to create a service and a deployment of a simple application.

```shell
basic-bundle
├── .imgpkg
│   ├── bundle.yml
│   └── images.yml
└── config.yml
```

You can push the above folder using the following command:

`imgpkg push -f examples/basic-bundle -b localhost:9001/simple-app-bundle`

Flags used in the command:

>   * `-b` indicates that the user want to push a bundle to the registry

Reviewer: ```suggestion
  * `-b` indicates that the user wants to push a bundle to the registry
```

>   * `-f` indicates the folder the user want to package

Reviewer: ```suggestion
  * `-f` indicates the folder the user wants to package
```

The output displays all the files that will be packaged and the destination of the bundle:
```shell
dir: .
dir: .imgpkg
file: .imgpkg/bundle.yml
file: .imgpkg/images.yml
file: config.yml
Pushed 'localhost:9001/simple-app-bundle@sha256:70225df0a05137ac385c95eb69f89ded3e7ef3a0c34db43d7274fd9eba3705bb'
Succeeded
```

#### Relocate to a different registry

This step will relocate the bundle configuration and the images associated with it.

To relocate, use the following command:
`imgpkg copy -b localhost:9001/simple-app-bundle --to-repo localhost:5000/simple-app-new-repo`

Flags used in the command:

>   * `-b` indicates that the user want to copy a bundle from the Registry

Reviewer: ```suggestion
  * `-b` indicates that the user wants to copy a bundle from the Registry
```
